pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,789 @@
+
+	if !hasOldEntry {
+		// create alias
+		_, err := p.kmsClient.CreateAlias(ctx, &kms.CreateAliasInput{
+			AliasName:   aws.String(entry.AliasName),
+			TargetKeyId: &entry.Arn,
+		})
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to create alias: %v", err)
+		}
+		p.log.Debug("Alias created", aliasNameTag, entry.AliasName, keyArnTag, entry.Arn)
+	} else {
+		// update alias
+		_, err := p.kmsClient.UpdateAlias(ctx, &kms.UpdateAliasInput{
+			AliasName:   aws.String(entry.AliasName),
+			TargetKeyId: &entry.Arn,
+		})
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to update alias: %v", err)
+		}
+		p.log.Debug("Alias updated", aliasNameTag, entry.AliasName, keyArnTag, entry.Arn)
+
+		select {
+		case p.scheduleDelete <- oldEntry.Arn:
+			p.log.Debug("Key enqueued for deletion", keyArnTag, oldEntry.Arn)
+		default:
+			p.log.Error("Failed to enqueue key for deletion", keyArnTag, oldEntry.Arn)
+		}
+	}
+	return nil
+}
+
+func (p *Plugin) setCache(keyEntries []*keyEntry) {
+	// clean previous cache
+	p.entries = make(map[string]keyEntry)
+
+	// add results to cache
+	for _, e := range keyEntries {
+		p.entries[e.PublicKey.Id] = *e
+		p.log.Debug("Key loaded", keyArnTag, e.Arn, aliasNameTag, e.AliasName)
+	}
+}
+
+// scheduleDeleteTask ia a long running task that deletes keys that were rotated
+func (p *Plugin) scheduleDeleteTask(ctx context.Context) {
+	backoffMin := 1 * time.Second
+	backoffMax := 60 * time.Second
+	backoff := backoffMin
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.log.Debug("Stopping schedule delete task", reasonTag, ctx.Err())
+			p.notifyDeleteDone()
+			return
+		case keyArn := <-p.scheduleDelete:
+			log := p.log.With(keyArnTag, keyArn)
+			_, err := p.kmsClient.ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
+				KeyId:               aws.String(keyArn),
+				PendingWindowInDays: aws.Int32(7),
+			})
+
+			if err == nil {
+				log.Debug("Key deleted")
+				backoff = backoffMin
+				p.notifyDeleteDone()
+				continue
+			}
+
+			var notFoundErr *types.NotFoundException
+			if errors.As(err, &notFoundErr) {
+				log.Error("No such key, dropping from delete schedule")
+				p.notifyDeleteDone()
+				continue
+			}
+
+			var invalidArnErr *types.InvalidArnException
+			if errors.As(err, &invalidArnErr) {
+				log.Error("Invalid ARN, dropping from delete schedule")
+				p.notifyDeleteDone()
+				continue
+			}
+
+			var invalidState *types.KMSInvalidStateException
+			if errors.As(err, &invalidState) {
+				log.Error("Key was on invalid state for deletion, dropping from delete schedule")
+				p.notifyDeleteDone()
+				continue
+			}
+
+			log.Error("It was not possible to schedule key for deletion", reasonTag, err)
+			select {
+			case p.scheduleDelete <- keyArn:
+				log.Debug("Key re-enqueued for deletion")
+			default:
+				log.Error("Failed to re-enqueue key for deletion")
+			}
+			p.notifyDeleteDone()
+			backoff = min(backoff*2, backoffMax)
+			time.Sleep(backoff)
```

Comment:
Using an actual sleep here is going to 1) keep this from responding to context cancellation in a timely manner, 2) make it harder to test...

Reply:
The backoff as applied also ends up applying the backoff to all deletion operations including those for unrelated keyArn's